design/secrets.md

```diff
@@ -0,0 +1,81 @@
+    By having the user name the secrets directly, it will be explicit and clearer for the user to understand how credentials are added and stored, and make it easier for them to understand and edit their own deployments.
+
+- See discussion https://github.com/vmware-tanzu/velero/pull/2259#discussion_r384700723 for the two items below.
+    - The `velero backup-location (create|set)` will have a new flag to set the credentials based on the option selected above. If the first option, the new flag will be `--credentials mapStringString` which sets the name of the corresponding credentials secret for a provider. Format is provider:credentials-secret-name.
+      If the second option, the new flag will be `--credentials string` where the argument will be the name of the credential set when adding a plugin.
+    - The `velero snapshot-location (create|set)` will have a new flag to set the credentials based on the option selected above. If the first option, the new flag will be `--credentials mapStringString` which sets the name of the corresponding credentials secret for a provider. Format is provider:credentials-secret-name.
+      If the second option, the new flag will be `--credentials string` where the argument will be the name of the credential set when adding a plugin.
+
+  Note that for this logic to work we must have a controller loop checking for when a corresponding secret is present before marking the BSL/VSL as ready.
+
+- The spec for BSLs and VSLs will updated to contain a new field `Credentials`. The value in this field will be the name of the key in the `cloud-credentials` secret to use for authenticating with the storage provider.
+
+- Plugins will need to be invoked differently so that the correct credential is used.
+  Currently, there is a single secret, which is mounted into every pod deployed by Velero (the Velero Deployment and the Restic DaemonSet).
+  This secret contains a single data key (`cloud`) which is accessible within the pods at the path `/credentials/cloud`.
+  This path is made known to all plugins through provider specific environment variables.
+  All possible provider environment variables are set to this path.
+  Instead of setting the environment within all the pods, we can modify `restartableProcess` to set the environment variables before running a plugin process.
+  Each plugin process would still have the same set of environment variables set, however the value used for each of these variables would instead be a different path within the `/credentials` directory, formed from the credential selected for a particular BSL/VSL.
+  Taking this approach would not require any changes from plugins as the credentials information would be made available to them in the same way.
+  We will also need to ensure that the restic controllers are updated in the same way so that correct credentials are used (when creating a `ResticRepository` or processing `PodVolumeBackup`/`PodVolumeRestore`).
+
+## Alternatives Considered
```

Comment:
Here is another alternate proposal:

Introduce a new `credentials` config in the `BackupStorageLocationSpec` and `VolumeSnapshotLocationSpec`.
This config will store a reference to a kubernetes secret that is managed is created by the user. using a command, say `velero credentials create` or at the time of creating the BSL or VSL.

The plugins will:
1. Read the `credentials` config key, if it exists will fetch the credentials from the secret.
1. Create a session.

We could also make `credentials` a field which is a kubernetes object reference to a secret. I am not sure what advantages it can get us but is something that can be thought through in more detail.

Pros:
1. Allows us to create credentials per BSL/VSL and not one per plugin. Users can add the, say, `velero-plugin-for-aws` once and use a different set of credentials based on what the BSL is configured to use. This, IMO, is better than allowing credentials per plugin.
1. Allows users to create separate credentials: one for accessing the object store, another for volume snapshotting. Following the security best practices of principle of least privilege.
1. This can be a step towards multi-tenancy, albeit a small one.
1. Users can also self-manage the BSL/VSL credentials when they wish to rotate credentials. Following the best practices of regular credential rotation.
1. Velero may never have to read the credentials. Somewhat reducing the security vulnerability surface.

Cons:
1. This will require plugin implementations to be updated to use the new behavior. However, existing plugins can still rely on the `cloud-credentials` secret to function as they currently do.

Reply:
> Allows users to create separate credentials

This is a great point.

Couple of thoughts just for the sake of reinforcing some ideas: with the credentials being bound to a provider, which is to say, bound to the plugin. it would be nice if we could somehow namespace the credentials on a per plugin basis. And, the BSL/VSL can reuse existing credentials, meaning, aws BSL X and aws BSL Y can both use credential A for aws.

Reply:
@ashish-amarnath your suggested approach was the path I initially went down, but was [concerned about the impact on existing plugins (first bullet point)](https://github.com/vmware-tanzu/velero/pull/2403#issuecomment-728132546). If we can leave the existing environment in place however, and we're comfortable with requiring plugins to make changes, then I think this is the much cleaner and simpler approach. Longer term, it would also allow us to move away from having provider specific information in our deployment.

One thing we would have to advertise though is that this would be supporting multiple credentials *but only for plugins that support it*. If we're okay with that, then I would much prefer to take this approach.

Reply:
@ashish-amarnath @carlisia I made a draft PR with the basic changes that would be needed to make the approach of specifying the credentials file in the config work: https://github.com/vmware-tanzu/velero-plugin-for-aws/pull/69
This change would obviously need more work (e.g. checking that the file exists) but this works with Velero as it is (if you update the secret and add the key to the BSL config). If we can agree that we're happy to put the change into the plugins, this would _much_ nicer :+1:

Reply:
@ashish-amarnath For your proposal, how would said credentials be exposed to restic, since it doesn't go through the plugin code path?

Reply:
One thing to consider with the above approach, I think it would be up to Velero to specify the value in the config rather than the user providing it. I think with this approach, the user would state which credential they want to use for that BSL (a new optional field) and then when calling `Init` with the config, Velero would pass through the config with the correct path for the credentials file (similarly to how we handle the `caCert` values).